### PR TITLE
Refactor walk tree

### PR DESCRIPTION
**Pull request by michaeldfallen**
This PR refactors the way we walk the tree to minimise the need to instantiate every step before walking the tree (dependency inversion principle) and prevent the walk tree function to need to know about each step type (Liskov substitution principle).

Rocking those SOLID principles.

### CODE REVIEW

**Comment by michaeldfallen** [comment]
Think this might be ready for the mainstream. @endamccormack can you review?

File: src/steps/check-your-answers/CheckYourAnswers.js [hunk 1]
```diff
@@ -1,32 +1,24 @@
...
-        .filter(step => step instanceof Question);
-    } else {
-      this.questions = [];
-    }
+  get middleware() {
+    return [this.collectSteps, ...super.middleware];
+  }
 
-    this.questions.forEach(step => this.waitFor(step.ready()));
+  collectSteps(req, res, next) {
```
Comment (endamccormack) [inline src/steps/check-your-answers/CheckYourAnswers.js]: nice

File: test/flow/journey.test.js [hunk 2]
```diff
@@ -35,7 +35,7 @@ const handlerTest = ({ test, options: extraOptions }) => {
 };
 
 
-describe('Journey', () => {
+describe('journey/journey', () => {
```
Comment (endamccormack) [inline test/flow/journey.test.js]: why did you make this change?

Comment (michaeldfallen) [reply test/flow/journey.test.js]: Journey moved in to it's own package... so I described where it was in the test suite. Now I realise that it's moved again. I should fix that

**Pull request merged**